File: python/examples/custom_observability/test_custom_scopes.py
"""Test: Verify with_custom_scopes includes scenario + DB spans."""
import asyncio
import os
import time

# Ensure no LangWatch data is sent
os.environ["LANGWATCH_API_KEY"] = ""

from opentelemetry import trace
from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter

import scenario
from scenario import (
    run,
    AgentRole,
    AgentAdapter,
    AgentInput,
    AgentReturnTypes,
    user,
    agent,
    succeed,
    with_custom_scopes,
)


def get_scope_name(span: ReadableSpan) -> str:
    scope = getattr(span, "instrumentation_scope", None) or getattr(
        span, "instrumentation_info", None
    )
    if scope:
        return getattr(scope, "name", "unknown")
    return "unknown"


memory_exporter = InMemorySpanExporter()

scenario.configure(
    observability={
        "span_filter": with_custom_scopes("my-app/database"),
        "span_processors": [SimpleSpanProcessor(memory_exporter)],
        "instrumentors": [],
    }
)

# Simulate DB spans (should be kept)
db_tracer = trace.get_tracer("my-app/database")
db_span = db_tracer.start_span("db.query SELECT * FROM users")
db_span.end()

# Simulate HTTP noise (should be filtered for LangWatch, but captured by memory exporter)
noise_tracer = trace.get_tracer("http-server")
noise_span = noise_tracer.start_span("GET /api/health")
noise_span.end()


class DummyUserAgent(AgentAdapter):
    role = AgentRole.USER

    async def call(self, input: AgentInput) -> AgentReturnTypes:
        return "unused"


class EchoAgent(AgentAdapter):
    role = AgentRole.AGENT

    async def call(self, input: AgentInput) -> AgentReturnTypes:
        # Simulate a DB call during agent execution
        tracer = trace.get_tracer("my-app/database")
        span = tracer.start_span("db.query INSERT INTO responses")
        span.end()
        last_msg = input.messages[-1] if input.messages else None
        content = last_msg.get("content", "") if last_msg else ""
        return f"Echo: {content}"


async def main():
    print("Running scenario with database-calling agent...")

    result = await run(
        name="custom-scopes-test",
        description="Test that with_custom_scopes includes scenario + DB spans",
        agents=[EchoAgent(), DummyUserAgent()],
        script=[user("Test with database spans"), agent(), succeed()],
    )

    print(f"\nScenario result: {'passed' if result.success else 'failed'}")

    time.sleep(0.5)

    spans = memory_exporter.get_finished_spans()
    print(f"\nTotal spans collected: {len(spans)}")

    scenario_spans = [s for s in spans if get_scope_name(s) == "langwatch"]
    db_spans = [s for s in spans if get_scope_name(s) == "my-app/database"]
    noise_spans = [s for s in spans if get_scope_name(s) == "http-server"]

    print(f"  Scenario spans: {len(scenario_spans)}")
    print(f"  Database spans: {len(db_spans)}")
    print(f"  HTTP noise spans: {len(noise_spans)}")

    if not result.success:
        print("\nFAIL: Scenario did not succeed")
        exit(1)

    if db_spans:
        print("\nDatabase spans captured:")
        for span in db_spans:
            print(f"  - {span.name}")

    print("\nPASS: with_custom_scopes includes both scenario and database spans")
    print("   LangWatch exporter would export scenario + my-app/database spans.")
    print("   HTTP noise spans are filtered out by the LangWatch exporter.")
    exit(0)


if __name__ == "__main__":
    asyncio.run(main())

File: javascript/examples/custom-observability/test-custom-scopes.ts
/**
 * Test: Verify withCustomScopes() -- capture scenario spans + custom tagged spans.
 *
 * Demonstrates the advanced use case where a user wants to include their own
 * instrumented code (e.g., database calls) alongside scenario spans.
 */
import { trace } from "@opentelemetry/api";
import {
  SimpleSpanProcessor,
  InMemorySpanExporter,
} from "@opentelemetry/sdk-trace-base";
import type { ReadableSpan } from "@opentelemetry/sdk-trace-base";
import {
  run,
  AgentRole,
  user,
  agent,
  succeed,
  setupScenarioTracing,
  withCustomScopes,
} from "@langwatch/scenario";

/**
 * Returns the instrumentation scope name for a span, handling both
 * OTel SDK v1 (instrumentationLibrary) and v2 (instrumentationScope).
 */
function getScopeName(span: ReadableSpan): string {
  const s = span as any;
  return (
    s.instrumentationScope?.name ??
    s.instrumentationLibrary?.name ??
    "unknown"
  );
}

// --- Step 1: Set up span collection ---
const memoryExporter = new InMemorySpanExporter();
const collectorProcessor = new SimpleSpanProcessor(memoryExporter);

setupScenarioTracing({
  instrumentations: [],
  spanProcessors: [collectorProcessor],
  langwatch: "disabled" as any,
});

// --- Step 2: Create a "database" tracer under a custom scope ---
const dbTracer = trace.getTracer("my-app/database");
const httpTracer = trace.getTracer("http-server");

// Simulate a database-backed agent
const dbAgent = {
  role: AgentRole.AGENT as const,
  call: async (input: any) => {
    // Simulate a database query (tagged with custom scope)
    return dbTracer.startActiveSpan(
      "db.query SELECT users",
      async (dbSpan) => {
        try {
          // Simulate some work
          await new Promise((resolve) => setTimeout(resolve, 10));
          dbSpan.setAttribute(
            "db.statement",
            "SELECT * FROM users WHERE id = ?"
          );
          dbSpan.setAttribute("db.system", "postgresql");
          return "Found user: Alice (from database)";
        } finally {
          dbSpan.end();
        }
      }
    );
  },
};

// A dummy user agent is required by the framework for the user() script step,
// even though the content is provided inline (the agent's call() is never invoked).
const dummyUserAgent = {
  role: AgentRole.USER as const,
  call: async () => "unused",
};

// --- Step 3: Create HTTP noise (should be excluded) ---
const httpSpan = httpTracer.startSpan("GET /api/health");
httpSpan.end();

// --- Step 4: Run scenario ---
console.log("Running scenario with database-backed agent...");

const result = await run({
  name: "custom-scopes-test",
  description: "Test that custom scopes are captured alongside scenario spans",
  agents: [dbAgent, dummyUserAgent],
  script: [user("Look up user Alice"), agent(), succeed()],
});

console.log(`\nScenario result: ${result.success ? "passed" : "failed"}`);

// --- Step 5: Inspect spans by scope ---
await new Promise((resolve) => setTimeout(resolve, 500));
await collectorProcessor.forceFlush();

const spans = memoryExporter.getFinishedSpans();
const byScope = new Map<string, typeof spans>();
for (const span of spans) {
  const scope = getScopeName(span);
  if (!byScope.has(scope)) byScope.set(scope, []);
  byScope.get(scope)!.push(span);
}

console.log(`\nTotal spans: ${spans.length}`);
console.log("\nSpans by instrumentation scope:");
for (const [scope, scopeSpans] of byScope) {
  console.log(`\n  ${scope} (${scopeSpans.length} spans):`);
  for (const span of scopeSpans) {
    const attrs = Object.entries(span.attributes)
      .filter(([k]) => k.startsWith("db."))
      .map(([k, v]) => `${k}=${v}`)
      .join(", ");
    console.log(`    - ${span.name}${attrs ? ` [${attrs}]` : ""}`);
  }
}

// --- Step 6: Show what withCustomScopes would filter ---
const filters = withCustomScopes("my-app/database");
console.log("\n--- Filter config for LangWatchTraceExporter ---");
console.log('withCustomScopes("my-app/database") would include:');
console.log(
  `  @langwatch/scenario spans (${byScope.get("@langwatch/scenario")?.length ?? 0})`
);
console.log(
  `  my-app/database spans (${byScope.get("my-app/database")?.length ?? 0})`
);
console.log(
  `  http-server spans (${byScope.get("http-server")?.length ?? 0}) -- filtered out`
);

const scenarioCount = byScope.get("@langwatch/scenario")?.length ?? 0;
const dbCount = byScope.get("my-app/database")?.length ?? 0;

if (scenarioCount > 0 && dbCount > 0) {
  console.log("\nPASS: Both scenario and custom-scoped spans are captured");
  console.log(
    "   Database spans appear alongside scenario spans for debugging."
  );
} else {
  console.error("\nFAIL: Expected both scenario and database spans");
  console.error(
    `   Scenario spans: ${scenarioCount}, Database spans: ${dbCount}`
  );
  process.exit(1);
}

process.exit(0);
